### PR TITLE
Fix/handle multiple classes in same files correctly

### PR DESCRIPTION
**Pull request by gro1m**
Hi @aconrad 
I try to continue on PR #113 here, as @smortvedt did not yet reply and may have lost interest in the project.

As a first, I just merged in his changes into my master and put it into a new branch.

### CODE REVIEW

**Comment by gro1m** [comment]
Hi @aconrad 
Also this is ready for review, unfortunately Python 3.8 and Python 3.9 are still red, albeit the diff is small, that might give an easier indication what is different in Python 3.8 and 3.9...

**Comment by aconrad** [comment]
> Hi @aconrad 
> Also this is ready for review, unfortunately Python 3.8 and Python 3.9 are still red, albeit the diff is small, that might give an easier indication what is different in Python 3.8 and 3.9...

Do you have both py37 and py38/py39 installed? The generated coverage files for each interpreters might be worth looking at. I'm not in front of my laptop but if you were to upload the files in a new GitHub issue that would be great so we can take a look at what is different. Do you think you can do that?

File: pycobertura/cobertura.py [hunk 1]
```diff
@@ -77,32 +77,50 @@ def _load_from_file(self, report_file):
     def _load_from_string(self, s):
         return ET.fromstring(s)
 
+    @memoize
+    def _get_class_elements_by_filename(self, filename):
+        syntax = f"./packages//class[@filename='{filename}']"
+        classes = []
+        for c in self.xml.xpath(syntax):
+            classes.append(c)
```
Comment (aconrad) [inline pycobertura/cobertura.py:85]: Can we change this code to be consistent with the variable names used in `_get_lines_by_filename()`?

```suggestion
        classElements = []
        for classElement in self.xml.xpath(syntax):
            classElements.append(classElement)
```

Alternatively, if `classElement` is too verbose, I'd be okay to use `classEl`.

File: pycobertura/cobertura.py [hunk 2]
```diff
@@ -77,32 +77,50 @@ def _load_from_file(self, report_file):
...
 
     def line_rate(self, filename=None):
         """
         Return the global line rate of the coverage report. If the
         `filename` file is given, return the line rate of the file.
         """
+        lineRate = 0
         if filename is None:
-            el = self.xml
+            lineRate = float(self.xml.get("line-rate"))
```
Comment (aconrad) [inline pycobertura/cobertura.py:114]: Can we early return here?

```suggestion
            return float(self.xml.get("line-rate"))
```

... and remove the `else` statement below.

File: pycobertura/cobertura.py [hunk 3]
```diff
@@ -114,27 +132,33 @@ def branch_rate(self, filename=None):
...
         """
-        el = self._get_class_element_by_filename(filename)
-        lines = el.xpath("./lines/line[@hits=0]")
-        return [int(line.attrib["number"]) for line in lines]
+        classes = self._get_class_elements_by_filename(filename)
+        return [
+            int(line.get("number"))
+            for cl in classes
+            for line in cl.xpath("./lines/line[@hits=0]")
+        ]
```
Comment (aconrad) [inline pycobertura/cobertura.py:148]: Same here, can we rename for consistency?

```suggestion
        classElements = self._get_class_elements_by_filename(filename)
        return [
            int(line.get("number"))
            for classElement in classElements
            for line in classElement.xpath("./lines/line[@hits=0]")
        ]
```

File: pycobertura/cobertura.py [hunk 4]
```diff
@@ -114,27 +132,33 @@ def branch_rate(self, filename=None):
...
         """
-        el = self._get_class_element_by_filename(filename)
-        lines = el.xpath("./lines/line[@hits>0]")
-        return [int(line.attrib["number"]) for line in lines]
+        classes = self._get_class_elements_by_filename(filename)
+        return [
+            int(line.get("number"))
+            for cl in classes
+            for line in cl.xpath("./lines/line[@hits>0]")
+        ]
```
Comment (aconrad) [inline pycobertura/cobertura.py:161]:
```suggestion
        classElements = self._get_class_elements_by_filename(filename)
        return [
            int(line.get("number"))
            for classElement in classElements
            for line in classElement.xpath("./lines/line[@hits>0]")
        ]
```

**Review by aconrad — APPROVED** [review]
Thanks for reviving this PR @gro1m, and @smortvedt for your original work! Sorry it took so long to review.

I'm glad we had a sample in our own test sample coverage.xml, I'm surprised I missed that. Good catch! 🎉

I suggested consistency for variable names across the changes but overall it looks good to me.

File: pycobertura/cobertura.py (hunk 4 above)
Comment (gro1m) [reply pycobertura/cobertura.py]: done in https://github.com/aconrad/pycobertura/pull/132/commits/45e1b7b103bd59c9ff4756375a69fbcd479d0608

File: pycobertura/cobertura.py (hunk 3 above)
Comment (gro1m) [reply pycobertura/cobertura.py]: done in https://github.com/aconrad/pycobertura/pull/132/commits/45e1b7b103bd59c9ff4756375a69fbcd479d0608

File: pycobertura/cobertura.py (hunk 2 above)
Comment (gro1m) [reply pycobertura/cobertura.py]: done in https://github.com/aconrad/pycobertura/pull/132/commits/45e1b7b103bd59c9ff4756375a69fbcd479d0608

File: pycobertura/cobertura.py (hunk 1 above)
Comment (gro1m) [reply pycobertura/cobertura.py]: done in https://github.com/aconrad/pycobertura/pull/132/commits/45e1b7b103bd59c9ff4756375a69fbcd479d0608

**Pull request merged**